### PR TITLE
Reworked wep pickup

### PR DESCRIPTION
**Pull request by TimGoll**
This is still WIP, there are a few issues left.

This reworked weapon pickup uses the PlayerCanPickupWeapon hook to solve the problem that values set to the weapon entity on runtime are deleted on pickup

### CODE REVIEW

**Comment by TimGoll** [comment]
This seems like a really hacky solution, but it does work. Please review!

File: gamemodes/terrortown/gamemode/server/sv_weapon_pickup.lua [hunk 1]
```diff
@@ -8,106 +8,58 @@ if not plymeta then
...
-		-- get droppable weapon from given slot
-		for i = 1, #weps do
-			throwWeapon = weps[i]
-
-			-- found a weapon that is allowed to be dropped
-			if IsValid(throwWeapon) and throwWeapon.AllowDrop then
-				break
-			end
-		end
+function GetBlockingWeapon(ply, wep)
```
Comment (Alf21) [inline gamemodes/terrortown/gamemode/server/sv_weapon_pickup.lua]: Why making this a public function and not a player's metatable function?

Comment (Alf21) [inline gamemodes/terrortown/gamemode/server/sv_weapon_pickup.lua]: Missing docu

File: gamemodes/terrortown/gamemode/server/sv_weapon_pickup.lua [hunk 2]
```diff
@@ -8,106 +8,58 @@ if not plymeta then
...
-
-	-- copy data from picked up weapon
-	if isfunction(newWep.SetClip1) then
-		newWep:SetClip1(clip1)
-	end
+	return throwWeapon, false
+end
 
-	self:GiveAmmo(ammo_num, ammo_type)
+function PrepareAndDropWeapon(ply, wep)
```
Comment (Alf21) [inline gamemodes/terrortown/gamemode/server/sv_weapon_pickup.lua]: Missing docu

File: gamemodes/terrortown/gamemode/server/sv_weaponry.lua [hunk 3]
```diff
@@ -627,7 +634,53 @@ function GM:WeaponEquip(wep, ply)
...
 		AddWeaponToInventoryAndNotifyClient(ply, wep)
+
+		-- there is a glitch that picking up a weapon does not refresh the weapon cache on
+		-- the client. Therefore the client has to be notified to updated its cache
+		net.Start("ttt2_switch_weapon_update_cache")
+		net.Send(ply)
 	end
+
+	-- handle all this stuff in the next frame since the owner is not yet valid
+	timer.Simple(0, function()
```
Comment (Alf21) [inline gamemodes/terrortown/gamemode/server/sv_weaponry.lua]: please localize this function

Comment (TimGoll) [reply gamemodes/terrortown/gamemode/server/sv_weaponry.lua]: I need `wep` and `ply`
all I could do would be to create a new function and call it inside this.

File: gamemodes/terrortown/gamemode/server/sv_player_ext.lua [hunk 4]
```diff
@@ -1037,8 +1038,78 @@ end
...
+
+	pFootPos.z = pFootPos.z + 32
+
+	wep:SetPos(pFootPos)
+
+	return wep
+end
+
+function plymeta:PickupWeaponClass(wepCls, dropBlockingWeapon)
+	local wep = ents.Create(wepCls)
```
[the remark this answers is not in the document]
Comment (TimGoll) [reply gamemodes/terrortown/gamemode/server/sv_player_ext.lua]: yes we have to check first, then drop the old weapon, then add the new weapon

File: gamemodes/terrortown/gamemode/server/sv_weaponry.lua (hunk 3 above)
Comment (Alf21) [reply gamemodes/terrortown/gamemode/server/sv_weaponry.lua]: you can localize this function right before

File: gamemodes/terrortown/gamemode/server/sv_player_ext.lua (hunk 4 above)
[the remark this answers is not in the document]
Comment (Alf21) [reply gamemodes/terrortown/gamemode/server/sv_player_ext.lua]: https://wiki.garrysmod.com/page/weapons/GetStored

File: gamemodes/terrortown/gamemode/server/sv_weaponry.lua [hunk 5]
```diff
@@ -644,6 +698,11 @@ end
 function GM:PlayerDroppedWeapon(ply, wep)
 	if IsValid(wep) and IsValid(ply) and wep.Kind then
```
Comment (Alf21) [inline gamemodes/terrortown/gamemode/server/sv_weaponry.lua]: you can minimize the scope

**Pull request merged**